### DISCUSSION TITLE
Add a warning for `time.sleep`

### DISCUSSION THREAD

**Discussion by encukou** [Textualize/textual#2843] [Ideas]
Hello,
Some time ago Will complained on Twitter that users [use `time.sleep` for experiments](https://twitter.com/willmcgugan/status/16628105528742215680). Let's complain when they do that!

To be clear: for finding blocking code in an asynchronous app, you should use a profiler, watchdog task, or similar. This issue is about adding a *hint* to people who don't understand or care about asynchrony, and just wanna write a TUI.

Unfortunately, most ways of detecting `time.sleep` are ugly: suggestions include monkey-patching `time` or analyzing AST .
Adding an [audit event](https://docs.python.org/3/library/audit_events.html) sounds like a plausible solution. I can add that to Python 3.13+, but I need cooperation from you. Specifically, I only want to add it if we can prove that it'll be useful to you.
The worry is about is performance. Audit events are cheap, but using a Python function as an audit hook might noticeably slow down *all* auditable events (not just sleep).

If it's too slow, it's possible to optimize -- but let's measure first. You're the best team to do that :)
Could you:

* Build Python with [this patch](https://github.com/encukou/cpython/pull/25/files) ([instructions here](https://devguide.python.org/getting-started/setup-building/), happy to help)
* Run Textual benchmarks (I assume you have some) with and without [somp audit hook for `sleep`](https://github.com/encukou/cpython/pull/25/files#diff-1cab14f7e86d7fddf1d9a2cb12c3f5ef50844cf932eb67a97f7fb6acff8330c2R520-R523) added at the start.
* Decide if the slowdown is worth it for you.

**Comment by willmcgugan** [comment]
Hi Petr. Thanks for looking in to this!

I'd imagined the audit hooks would only be enabled in debug mode. That way we could inform devs that they may be doing something wrong. The audit hooks would generally be disabled when the app is shipped.

I'm sure we would make use of the audit hook if it was available. Any slowdown is probably not going to be much of an issue for us...

**Comment by encukou** [comment]
Python 3.13.0alpha 1 is now out and it emits an auditing event for `time.sleep`.
Python tests have some [example usage](https://github.com/python/cpython/blob/4a929d432b48775096d40f5c72bcd9be052b0a2d/Lib/test/audit-tests.py#L521-L531).

**Reply by willmcgugan** [reply]
That's awesome. Thanks for doing this!
